share/eds/p2p/client.go

```diff
@@ -0,0 +1,161 @@
+	}
+}
+
+// RequestEDS requests the full EDS from one of the given peers.
+//
+// The peers are requested in a round-robin manner with retries until one of them gives a valid
+// response. Blocks forever until the context is canceled or a valid response is given.
+func (c *Client) RequestEDS(
+	ctx context.Context,
+	dataHash share.DataHash,
+	peers peer.IDSlice,
+) (*rsmt2d.ExtendedDataSquare, error) {
+	req := &p2p_pb.EDSRequest{Hash: dataHash}
+
+	// requests are retried for every peer until a valid response is received
+	excludedPeers := make(map[peer.ID]struct{})
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+		}
+
+		// if no peers are left, return
+		if len(peers) == len(excludedPeers) {
+			return nil, errNoMorePeers
+		}
+
+		for _, to := range peers {
+			select {
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			default:
+			}
+
+			// skip over excluded peers
+			if _, ok := excludedPeers[to]; ok {
+				continue
+			}
+
+			eds, err := c.doRequest(ctx, req, dataHash, to)
+			if err != nil {
+				// peer has misbehaved, exclude them from round-robin
+				excludedPeers[to] = struct{}{}
+				log.Errorw("client: eds request to peer failed", "peer", to, "hash", dataHash.String())
+			}
+
+			// eds is nil when the request was valid but couldn't be served
+			if eds != nil {
+				return eds, nil
+			}
+		}
+	}
+}
+
+func (c *Client) doRequest(
+	ctx context.Context,
+	req *p2p_pb.EDSRequest,
+	dataHash share.DataHash,
+	to peer.ID,
+) (*rsmt2d.ExtendedDataSquare, error) {
+	log.Debugf("client: requesting eds %s from peer %s", dataHash.String(), to)
+	stream, err := c.host.NewStream(ctx, to, c.protocolID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open stream: %w", err)
+	}
+
+	// request status from peer to provide the requested EDS
+	err = stream.SetWriteDeadline(time.Now().Add(writeDeadline))
+	if err != nil {
+		log.Warn(err)
+	}
+	_, err = serde.Write(stream, req)
+	if err != nil {
+		stream.Reset() //nolint:errcheck
+		return nil, fmt.Errorf("failed to write request to stream: %w", err)
+	}
+	err = stream.CloseWrite()
+	if err != nil {
+		return nil, fmt.Errorf("failed to close write on stream: %w", err)
+	}
+
+	// read and parse status from peer
+	resp := new(p2p_pb.EDSResponse)
+	err = stream.SetReadDeadline(time.Now().Add(readDeadline))
+	if err != nil {
+		log.Warn(err)
+	}
+	_, err = serde.Read(stream, resp)
+	if err != nil && err != io.EOF {
+		stream.Reset() //nolint:errcheck
+		return nil, fmt.Errorf("failed to read status from stream: %w", err)
+	}
+
+	switch resp.Status {
+	case p2p_pb.Status_OK:
+		// read all CARv1 header and ODS bytes from stream
+		carBytes, err := io.ReadAll(stream)
+		if err != nil && err != io.EOF {
+			return nil, fmt.Errorf("unexpected error while reading ods from stream: %w", err)
+		}
+		// use header and ODS bytes to construct EDS and verify it against dataHash
+		eds, err := eds.ReadEDS(ctx, bytes.NewReader(carBytes), dataHash)
```

Comment:
We can give the stream directly to the ReadEDS.

Reply:
sick, dont know why i didnt think about trying that